### PR TITLE
GitHub Actions Sandbox

### PR DESCRIPTION
**Pull request by Tyler-g-hudson**
This PR contains three changes to the GitHub repository:

1. It implements workflow which contains a matrix of five GitHub Actions runs - Linux Latest Release, Linux Latest Debug, Linux Minimum Release, macOS Latest Release, and macOS Latest Debug. These each build the conda environment, build the repository, and run ctest on it with a varying set of OS, dependencies, and other options as well as excluding certain known problem tests.
2, 3. A ".github/conda_envs/" folder has been added to the repo which contains two environment files, one which contains the typical conda environment information and one of which describes a more restricted minimum requirements environment.

### CODE REVIEW

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Can you add `pull_request` here as well?

Ideally, the CI build should be triggered whenever a PR is first created and whenever the PR is updated (by the PR author pushing changes to the PR branch). Right now, a CI build will only be triggered by the latter. (It didn't trigger when you created this PR, for example.)

```suggestion
on: [pull_request, push]
```

File: env/environment.yml
Comment (gmgunter) [inline env/environment.yml]: Initially I set this upper bound on the HDF5 version, because my experience is that the HDF5 folks don't follow [semantic versioning](https://semver.org/) and tend to introduce a lot of breaking changes even in their patch releases. So when I'm showing people how to build isce3, I try to avoid building with a newer version of HDF5 than I personally have tested with, in case some new breaking changes have been introduced.

But it occurs to me now that we should remove this upper bound for purposes of CI testing for basically the same reason -- if a new breaking change is introduced in a recent HDF5 version, it would be good for the CI system to let us know early on so we can address it quickly.

```suggestion
  - hdf5>=1.10.2
```

File: env/minimum-reqs.yml
Comment (gmgunter) [inline env/minimum-reqs.yml]: I think we can remove the comments like this, since they're not providing much useful information now that we've got a working minimum environment.

The `Current ver: 3.4.0` is not super useful because it's specific to a particular platform at a particular point in time. This version could change without warning as new packages versions are added to conda.

The `EIGEN: Important` part was useful initially to indicate which packages we thought might be most likely to introduce breakages in isce3 upon version changes (so we could try harder to pin those at their exact minimum supported version and be a little more flexible about the other packages). Now that we've got a minimum-reqs environment, though, I'm not sure that we need to persist this info in a comment.

There's not really a strict distinction between an "Important" dependency and an unimportant one. This was just my intuition about which dependencies we should prioritize.

File: env/environment.yml
Comment (gmgunter) [inline env/environment.yml]: Let's just name it `isce3`

```suggestion
name: isce3
```

File: env/minimum-reqs.yml
Comment (gmgunter) [inline env/minimum-reqs.yml]:
```suggestion
name: isce3
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]:
```suggestion
          environment-name: isce3
```

**Review by gmgunter — COMMENTED** [review]
Nice work @Tyler-g-hudson! Here are a few quick notes on things that I missed in our earlier offline discussions

File: env/environment.yml
Comment (gmgunter) [inline env/environment.yml]: I think I'd prefer that we don't create a new top-level directory to store these environment YAML files. My preference would be to try to keep the root directory of the repository as simple as possible and avoid adding too much clutter (although admittedly we're not currently doing a great job of that in isce3 :grin:).

Instead, I think it might be better to either put these in the `.github/` directory (perhaps in `.github/envs/`?) so that they live near the workflow that uses them, or perhaps put them somewhere in the `tools/`, where most of the existing CI scripts & resource files live.

I think I'm slightly more in favor of putting them in `.github/envs/` than in `tools/`, but I'm also slightly apprehensive that the GitHub folks might decide to standardize that directory name for something else at some point in the future (like they do with `.github/workflows/` and `.github/actions/`), which could possibly create issues.

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: `ctest -E` uses a regex pattern to identify tests to exclude, so a`.` like this will match any character. If we want to match a literal `.` instead, we should escape the dot with a backslash, like this:

```suggestion
      ALL_EXCLUDE: stage_dem|ionosphere|background|pybind\.unwrap\.phass
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]:
```suggestion
            ctest_exclude: -E ".*($ALL_EXCLUDE|raster|unwrap\.snaphu)"
```

Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: There's actually a few different tests with "raster" in their name, and I think I only expect one of them to fail on macOS.

Let's see if we can make this pattern a little more specific in order to avoid unnecessarily excluding some tests

```suggestion
            ctest_exclude: -E ".*($ALL_EXCLUDE|io\.gdal\.raster|unwrap.snaphu)"
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Should all of these tests really be in `ALL_EXCLUDE`?

My impression was that the `background` and `pybind.unwrap.phass` test only failed on macOS.

And I thought the `ionosphere` tests were only failing on the Linux Minimum job.

If that's the case, let's see if we can move these excludes to the build matrix below so they're only excluded from specific jobs where they're expected to fail instead of being excluded from all workflows, as currently implemented.

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: >`test -d "$ISCE3_PREFIX/lib64"`

I think this line isn't actually working the way we want it to, because we haven't yet installed isce3.

Our intention here is to check whether the isce3 libraries were installed in `$ISCE3_PREFIX/lib64` or `$ISCE3_PREFIX/lib` (because the install location differs on different linux distributions). We're doing this by testing whether the directory `$ISCE3_PREFIX/lib64` exists.

But we haven't yet installed isce3 at this point, so that directory won't exist yet in either case. Instead, we should perform this check after the installation step (but before the `ctest` step). At that point, the directory containing the installed libraries (either `lib` or `lib64`) should exist.

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: I don't think it should be necessary to add the `ld` linker itself to `LD_LIBRARY_PATH`

```suggestion
            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
          else
            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Is it necessary to both `export LD_LIBRARY_PATH` \*and\* add `LD_LIBRARY_PATH` to the conda environment using `conda env config vars set`?

I think we could probably skip the first step and just add it to the conda environment

```suggestion
            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
          else
            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
          fi
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Just a small stylistic nitpick. Elsewhere, when we're evaluating environment variables, we just use `$VAR` (without curly braces) instead of `${VAR}`. 

This is the only exception to that here AFAICT, so let's ditch the braces for aesthetic reasons. They shouldn't be necessary in this case.

```suggestion
          export ISCE3_PREFIX=$GITHUB_WORKSPACE/install
```

**Review by gmgunter — COMMENTED** [review]
Few more suggestions :slightly_smiling_face:

File: .github/workflows/build-and-run.yml
Comment (Tyler-g-hudson) [reply .github/workflows/build-and-run.yml]: The reason I'm doing these export-and-then-add-to-environment things is to prevent the runner from having to deactivate and reactivate the environment. `conda env config vars set` requires the environment to be deactivated and reactivated before the variable becomes available for use - so if I need the variable later (e.g. using $ISCE3_PREFIX to set $LD_LIBRARY_PATH and then using $LD_LIBRARY_PATH to set $DYLD_LIBRARY_PATH) I'll have to deactivate the environment and reactivate it which requires some extra steps and and a couple of extra seconds of runtime.

Using `export` first sets the environment variable and makes it available for use within this step and requiring no deactivation and reactivation, and then using `conda env config vars set` makes it available for use in the following steps of the workflow. I've found this to be a faster and simpler way of doing the job.

Another alternative is also for me to just set these variables without referencing each other at all, which I believe would require an additional if/else statement and may make the script a little slower to modify but would delete these export/conda env config vars set steps:

```Shell
conda env config vars set ISCE3_PREFIX=${GITHUB_WORKSPACE}/install
conda env config vars set PYTHONPATH=$PYTHONPATH:${GITHUB_WORKSPACE}/install/packages
if test -d "${GITHUB_WORKSPACE}/install/lib64"; then
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib64
  if [ "$RUNNER_OS" == "macOS" ]; then
    conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib64
  fi
else
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib
  if [ "$RUNNER_OS" == "macOS" ]; then
    conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib
  fi
fi
```

Comment (Tyler-g-hudson) [reply .github/workflows/build-and-run.yml]: Scratch this! With the above comment about moving LD_LIBRARY_PATH setting to after `cmake --build build --target install` has been run, I don't end up having much of a reason to do these export statements anyway.

File: .github/workflows/build-and-run.yml
Comment (Tyler-g-hudson) [reply .github/workflows/build-and-run.yml]: My thinking here was that we were concerned about the nature of these tests generally, i.e. that they were written in such a way that they had greater potential to fail across a variety of platforms and dependencies (that is to say, if we were to add more tests to this CI workflow, all of the tests here had either a greater potential to fail or were considered otherwise flawed)

With the exception of stage_dem of course, which simply has a permissions problem and will be tested elsewhere.

If you'd like, I can move them though!

File: env/environment.yml
Comment (Tyler-g-hudson) [reply env/environment.yml]: Thoughts on using something less likely to be overwritten, like `.github/conda_envs/`?

Comment (gmgunter) [reply env/environment.yml]: Yeah, good idea, `.github/conda_envs/` sounds good to me :+1:

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [reply .github/workflows/build-and-run.yml]: Oh I see the reasoning now, thanks.

It's a fair point, but I'm thinking that it'd be better to try and maximize coverage of each job, and only add a test to `ALL_EXCLUDE` if it's been demonstrated to fail across multiple platforms and/or sets of dependencies. It doesn't cost much to move a test to `ALL_EXCLUDE` in the future. 

It seems like the `background` test is failing on macOS because we made some assumptions about runtime that aren't satisfied on the slower GHA macOS runners, and the `ionosphere` tests seem to be failing in the Linux Minimum job due to some issue with older versions of GDAL (though I'm not sure exactly what). So, at least in those two cases, we have some mechanistic reason to believe the tests should fail only in specific jobs.

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [reply .github/workflows/build-and-run.yml]: Oh right gotcha -- you needed to export `LD_LIBRARY_PATH` as well because you were using it to populate `DYLD_LIBRARY_PATH` as well in that same workflow step. Sorry, I forgot about that part when I made the initial comment but it makes sense now.

**Comment by gmgunter** [comment]
I'm not sure why we're not seeing the CI checks here on the PR page. I see it on @Tyler-g-hudson's fork here - https://github.com/Tyler-g-hudson/isce3-ci-sandbox/actions - but normally we'd see the pass/fail results on the PR itself.

Maybe the `.github/workflows/*.yml` file needs to already exist when the PR is created for this to work? Or maybe it's because the initial PR creation didn't trigger a CI workflow?

Either way, I suppose it's probably not a big deal. We can just check that the workflows on the fork are passing for now. After this is merged, I expect that we'll see CI checks on subsequent PRs, but we can verify that when we get to that point.

**Comment by rtburns-jpl** [comment]
Yeah I think that's a security feature, so we can inspect the workflow runs before they run under the organization. Once it's merged they'll run here as you expect.

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Let's see if we can [upgrade to `checkout@v3`](https://github.com/actions/checkout#whats-new). I see a few warnings on our [CI runs](https://github.com/Tyler-g-hudson/isce3-ci-sandbox/actions/runs/4167096208) currently that seem to be related to using the older version:

Screenshot from 2023-02-14 21-21-40

```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: According to [this](https://stackoverflow.com/questions/3146274/is-it-ok-to-use-dyld-library-path-on-mac-os-x-and-whats-the-dynamic-library-s), `DYLD_LIBRARY_PATH` on macOS behaves the same as `LD_LIBRARY_PATH` on Linux. So, on macOS, I think we \*only\* need to define `DYLD_LIBRARY_PATH` -- not both `LD_LIBRARY_PATH` \*and\* `DYLD_LIBRARY_PATH`.

Also, the `lib64` vs `lib` directory name ambiguity is really only an issue on Linux. According to [this](https://stackoverflow.com/a/67720256), `lib64` is used only on non-debian-based 64-bit Linux systems. On macOS, the library install directory will always be called `lib` -- never `lib64`.

With this in mind, I think we could try to simplify this logic a bit more. WDYT about this approach?

```
# Append to DYLD_LIBRARY_PATH on macOS or to LD_LIBRARY_PATH on Linux.
# Depending on the Linux flavor, the install dir name could be 'lib64' or 'lib'.
if [ "$RUNNER_OS" == "macOS" ]; then
  conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$ISCE3_PREFIX/lib
elif test -d "$ISCE3_PREFIX/lib64"; then
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
else
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
fi
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Another very minor pedantic comment --

In general in this file, you use a single blank line to separate blocks of code. But here you're using 2 blank lines. I think it would look a little cleaner if we tried to consistently use the same number of empty lines to space things out

```suggestion
      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass

    defaults:
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml]: Same here

```suggestion
          conda env config vars set PYTHONPATH=$PYTHONPATH:$GITHUB_WORKSPACE/install/packages

      # Get cmake command configured with relevant build information
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml:12]: Last but not least, let's add a comment here to explain why we're excluding these particular tests, because otherwise I will 100% forget why we did things this way :grin:

```suggestion
      # stage_dem requires AWS permissions to access the S3 bucket containing the DEM data.
      # pybind.unwrap.phass occasionally segfaults on GHA runners.
      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml:29]:
```suggestion
            # io.gdal.raster and unwrap.snaphu tests fail due to known unresolved issues
            # with GDAL memory mapping on macOS.
            # background reader/writer test sometimes fails on macOS due to slow runtimes.
            ctest_exclude: -E ".*($ALL_EXCLUDE|io\.gdal\.raster|unwrap\.snaphu|background)"
```

File: .github/workflows/build-and-run.yml
Comment (gmgunter) [inline .github/workflows/build-and-run.yml:42]:
```suggestion
              # ionosphere tests fail on Linux Minimum due to apparent issues with older GDAL versions
              ctest_exclude: -E ".*($ALL_EXCLUDE|ionosphere)"
```

File: .github/workflows/build-and-run.yml
Comment (Tyler-g-hudson) [reply .github/workflows/build-and-run.yml]: Looks good! I always love cutting a layer out of my conditional logic. I'll give it a try and see if it runs!

**Review by gmgunter — DISMISSED** [review]
LGTM, thanks!

**Comment by gmgunter** [comment]
@rtburns-jpl we talked over this PR together in our devops meeting a few weeks ago. Do you have any other comments that you want to make here? If not, can you go ahead and approve when you get a chance?

After this is merged, I was planning to ask @Tyler-g-hudson to look into replacing the environment.yml files with requirements.txt files, since we discussed wanting to minimize the number of redundant places in the codebase where isce3's requirements are enumerated.

**Review by rtburns-jpl — APPROVED** [review]
Nope nothing to add, all looking great to me.

**Review by gmgunter — APPROVED** [review]

**Pull request merged**
